### ISSUE TITLE
speed

### ISSUE THREAD

**Issue by mwufi** [PolicyEngine/policyengine#816]
Let's see what we can do to make it faster!!

Current status: (40s)
<img width="477" alt="image" src="https://user-images.githubusercontent.com/30219253/178111009-1f74e446-8237-4c44-abaf-d7c41b6075d8.png">

**Comment by mwufi** [comment]
Nothing really stands out, tbh

<img width="1451" alt="image" src="https://user-images.githubusercontent.com/30219253/178113766-63e033fb-4b0d-4e36-af1a-8d4897d240f6.png">

**Comment by nikhilwoodruff** [comment]
Yeah, unfortunately the microsimulation model takes a while (and it'll get longer), and is non-trivial to improve because there is a lot of complexity in the computation tree that needs to happen. That said, I think that the `apply_reform` call might be a good point of attack: I think it should be possible to get that down to under ~0.1s (it basically just modifies a set of existing variables in the OpenFisca model, and doesn't touch any large arrays like the rest of the simulation calculations).

The other good thing here might not be to increase speed, but make the wait less of a pain. An idea we've discussed at length is moving to a more rigorous task scheduling system (this is probably also safer/less risk of overloading the server), and enabling progress bars to be updated on the client side at various checkpoints in this stack trace.

**Comment by mwufi** [comment]
Oh! Thanks for the tip!! Yeah, it does seem like a lot of the computation tree is in `openfisca-core` and `openfisca-tools`... I'm actually really curious what the size of the computation that goes on under the hood is! Do you know how to peek into that?

**Comment by MaxGhenis** [comment]
@mwufi The computations occur in the [openfisca-us](https://github.com/PolicyEngine/openfisca-us) and [openfisca-uk](https://github.com/PolicyEngine/openfisca-uk) repos. [This notebook](https://colab.research.google.com/drive/1KJzkhuFO2h4YEbBKEDn_Pw22Q22_AzHe) shows a sample computation tree. I think the most expensive parts are looking up parameters, e.g. ACA subsidy parameters by ZIP code, and doing so for the ~100k households in the Current Population Survey (it's vectorized with numpy, but still lots of intertwining computations).

**Comment by nikhilwoodruff** [comment]
> Oh! Thanks for the tip!! Yeah, it does seem like a lot of the computation tree is in `openfisca-core` and `openfisca-tools`... I'm actually really curious what the size of the computation that goes on under the hood is! Do you know how to peek into that?
> 
> Happy to discuss anything of the "make the wait less of a pain" sort! Scheduling, UI, caching, etc...

Cool - OK, so here's a few ideas, any comments/thoughts welcome!

## 1. Status updates

Right now the client makes a request by clicking 'simulate', that request is POSTed to the server, and the server just gets straight to it and returns the result. Maybe a better system would use a ticketing system: the client POSTs the request, and gets the ticket number, the server adds it to a queue and keeps track of the progress of the simulation. The client can query the ticket number at any time to get the progress, and then when ready, specific chart outputs or the whole simulation result set to show on the client app. 

## 2. Progress bars

Once we've got the above system in place, it's probably a lot easier to give a progress bar UI that fills up when you're waiting for the simulation to complete. We could also use analytics from recent simulations to dynamically estimate the amount of time each simulation takes.

On your point about caching, we actually already cache simulations (for each version, new PolicyEngine deployments recalculate reforms). For example, [I just ran this reform](https://policyengine.org/uk/population-impact?basic_rate=25) and if you follow that link, it should appear within about a second because it's cached, compared to the usual 10-15 seconds for the UK sims.

**Comment by MaxGhenis** [comment]
We've also discussed enhancing openfisca-core to pass a baseline simulation, detect what needs to be recomputed based on the computation tree, and leave the rest as-is. Currently it recomputes everything from scratch.

If we could speed up computations, we could estimate impacts as users define the reform. For example, the four numbers at the top of the Population Impact page could be shown all the time, and then users could click them to see the rest of the impacts. We could mirror that experience on the Household Impact page, too; we could probably do that sooner given it's faster, but it'd be nice to maintain the parallel structure between Household and Population. I think these changes would significantly improve engagement.

**Comment by mwufi** [comment]
Ah! I need to get back to this! Last week I did start to figure out how the computation tree was built... via debugger...

Is there an easy entry point to look at, if we want to diff two computation trees? (say, baseline vs changed?)

**Comment by mwufi** [comment]
Wow, the computation tree is much bigger than I expected (over 1k lines!). Is this intended? https://gist.github.com/mwufi/edb108df706cf5a9298f3b672c64bb4c

**Comment by nikhilwoodruff** [comment]
Aha yeah I'm not surprised! It's a big model- though there are less variables than lines there due to caching (e.g. SNAP has its full tree shown the first time it's calculated, but not for other times it's used as an input to a variable).
